test/trigger/test_trigger.js

```diff
@@ -108,6 +108,64 @@ let testUnnecessaryGroupRegex = function(file, contents) {
   }
 };
 
+let testInvalidCapturingGroupRegex = function(file, contents) {
+  let json = eval(contents);
+
+  for (let i in json[0].triggers) {
+    let currentTrigger = json[0].triggers[i];
+
+    let containsMatches = false;
+    let triggerFunctions = [
+      'alarmText',
+      'alertText',
+      'condition',
+      'delaySeconds',
+      'infoText',
+      'preRun',
+      'run',
+      'suppressSeconds',
+      'tts',
+    ];
+
+    for (let j = 0; !containsMatches && j < triggerFunctions.length; j++) {
+      let currentTriggerFunction = currentTrigger[triggerFunctions[j]];
+      if (typeof currentTriggerFunction !== 'undefined' && currentTriggerFunction !== null)
+        containsMatches = currentTriggerFunction.toString().includes('matches');
+    }
+
+    let regexContainsCapturingGroups = /(\((?!\?:).*?\))/.test(currentTrigger.regex);
```

Comment:
*skeptical face*

I mean, I agree that this works for all triggers, but maybe you need a negative lookahead so that something like `\(real parens\)` doesn't match here too.

I actually wonder if the fact that javascript will yell at you for an unterminated group if all you need to do is verify that there's the beginning of a matching group?

Reply:
Interesting... I thought I had solved exactly this, and I did [on this line](https://github.com/quisquous/cactbot/pull/678/files#diff-7e6ad728e47e4d5e6d9d7f6cfe5a7dd6R150). Time for variable.

Reply:
I've updated this, but we should discuss if it's good enough. For instance, it won't match on `\(hey\)` anymore, but it will match on `(hey)` and `(hey\)`.

Reply:
If you just want to know how many captures the regex has, you can do something like this:
```javascript
let captures = new RegExp('(?:' + currentTrigger.regex.toString() + ')?').exec('').length - 1
```

Reply:
Ooh, that's nice.  :)

Reply:
Nice! This even found a case where there was two capturing groups for one language and one for the others.